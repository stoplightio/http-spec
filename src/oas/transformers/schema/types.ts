import type { JSONSchema7 } from 'json-schema';
import type { SchemaObject } from 'openapi3-ts/src/model/OpenApi';
import type { Schema } from 'swagger-schema-official';

export type OASSchemaObject = SchemaObject | Schema;

export type Converter = (schema: OASSchemaObject | JSONSchema7) => void;
