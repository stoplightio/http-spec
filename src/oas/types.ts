import { DeepPartial, IHttpOperation, IHttpService } from '@stoplight/types';
import { OpenAPIObject } from 'openapi3-ts';
import { Spec } from 'swagger-schema-official';

export const enum OasVersion {
  OAS2 = 2,
  OAS3 = 3,
}

/**
 * Service
 */

export interface ITransformServiceOpts<T> {
  document: DeepPartial<T>;
}

export type HttpServiceTransformer<T> = (opts: T) => IHttpService;

export type Oas2TransformServiceOpts = ITransformServiceOpts<Spec>;
export type Oas3TransformServiceOpts = ITransformServiceOpts<OpenAPIObject>;
export type Oas2HttpServiceTransformer = HttpServiceTransformer<Oas2TransformServiceOpts>;
export type Oas3HttpServiceTransformer = HttpServiceTransformer<Oas3TransformServiceOpts>;

/**
 * Operation
 */

export interface ITransformOperationOpts<T> {
  document: DeepPartial<T>;
  path: string;
  method: string;
}

export type HttpOperationTransformer<T> = (opts: T) => IHttpOperation;

export type Oas2TransformOperationOpts = ITransformOperationOpts<Spec>;
export type Oas3TransformOperationOpts = ITransformOperationOpts<OpenAPIObject>;
export type Oas2HttpOperationTransformer = HttpOperationTransformer<Oas2TransformOperationOpts>;
export type Oas3HttpOperationTransformer = HttpOperationTransformer<Oas3TransformOperationOpts>;
