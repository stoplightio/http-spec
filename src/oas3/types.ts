import { DeepPartial } from '@stoplight/types';
import { OpenAPIObject as _OpenAPIObject } from 'openapi3-ts';
import { PathItemObject } from 'openapi3-ts/src/model/OpenApi';
import { ISpecificationExtension } from 'openapi3-ts/src/model/SpecificationExtension';

import { TranslateFunction } from '../types';

export type Oas3TranslateFunction<P extends unknown[], R extends unknown = unknown> = TranslateFunction<
  DeepPartial<OpenAPIObject>,
  P,
  R
>;

export interface OpenAPIObject extends _OpenAPIObject {
  webhooks?: WebhooksObject;
}

export interface WebhooksObject extends ISpecificationExtension {
  [name: string]: PathItemObject;
}
