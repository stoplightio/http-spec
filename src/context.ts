import { TransformerContext } from './types';
import { maybeResolveLocalRef } from './utils';

export function createContext<T extends Record<string, unknown>>(document: T): TransformerContext<T> {
  return {
    document,
    maybeResolveLocalRef: maybeResolveLocalRef.bind(null, document),
  };
}
