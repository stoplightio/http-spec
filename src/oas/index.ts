export * from './operation';
export * from './types';
