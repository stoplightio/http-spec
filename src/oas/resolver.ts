import type { JsonPath } from '@stoplight/types';

import type { AvailableContext, RefResolver } from '../types';
import { resolveInlineRef } from './resolveInlineRef';

function inferContext(path: JsonPath): AvailableContext {
  if (path.length < 2 || path[0] !== 'paths') return 'service';
  if (path.length === 2 || path[3] === 'parameters' || path[3] === 'servers') return 'path';
  return 'operation';
}

const SHARED_COMPONENTS_KEYS = new WeakMap();

export function getSharedKey(value: object) {
  return SHARED_COMPONENTS_KEYS.get(value);
}

export const resolveRef: RefResolver = function (target) {
  const location: string[] = [];
  const resolved = resolveInlineRef(this.document, target.$ref, location);

  const context = inferContext(location);
  if (context !== null && this.context !== context) {
    this.context = context;
  }

  if (typeof resolved === 'object' && resolved !== null && context === 'service') {
    SHARED_COMPONENTS_KEYS.set(resolved, location[0] === 'components' ? location[2] : location[1]);
  }

  return resolved;
};
