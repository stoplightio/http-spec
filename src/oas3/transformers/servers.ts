import type { INodeVariable, IServer, Optional } from '@stoplight/types';
import pickBy = require('lodash.pickby');

import { withContext } from '../../context';
import { isNonNullable, isString } from '../../guards';
import { ArrayCallbackParameters } from '../../types';
import { entries } from '../../utils';
import { isServerObject, isServerVariableObject } from '../guards';
import { Oas3TranslateFunction } from '../types';

export const translateToServers = withContext<
  Oas3TranslateFunction<[path: Record<string, unknown>, operation: Record<string, unknown>], IServer[]>
>(function (path, operation) {
  let servers;
  if (Array.isArray(operation.servers)) {
    servers = operation.servers;
  } else if (Array.isArray(path.servers)) {
    servers = path.servers;
    this.context = 'path';
  } else if (Array.isArray(this.document.servers)) {
    servers = this.document.servers;
    this.context = 'service';
  } else {
    return [];
  }

  return servers.map(translateToServer, this).filter(isNonNullable);
});

export const translateToServer = withContext<
  Oas3TranslateFunction<ArrayCallbackParameters<unknown>, Optional<IServer>>
>(function (server) {
  if (!isServerObject(server)) return;

  const variables = translateServerVariables.call(this, server.variables);

  return {
    id: this.generateId(`http_server-${this.parentId}-${server.url}`),
    url: server.url,

    ...pickBy(
      {
        name: this.document.info?.title,
        description: server.description,
      },
      isString,
    ),

    ...pickBy(
      {
        variables,
      },
      isNonNullable,
    ),
  };
});

export const translateServerVariables: Oas3TranslateFunction<
  [variables: unknown],
  Optional<Record<string, INodeVariable>>
> = variables => {
  const serverVariables = entries(variables).map(translateServerVariable).filter(isNonNullable);
  return serverVariables.length > 0 ? Object.fromEntries(serverVariables) : undefined;
};

const translateServerVariable: Oas3TranslateFunction<
  ArrayCallbackParameters<[name: string, variable: unknown]>,
  Optional<[string, INodeVariable]>
> = function ([name, variable]) {
  if (!isServerVariableObject(variable)) return;

  return [
    name,
    {
      default: String(variable.default),

      ...pickBy(
        {
          description: variable.description,
        },
        isString,
      ),

      ...pickBy(
        {
          enum: Array.isArray(variable.enum) ? variable.enum.map(String) : undefined,
        },
        isNonNullable,
      ),
    },
  ];
};
