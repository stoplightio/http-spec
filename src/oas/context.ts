// @ts-expect-error: no types
import * as fnv from 'fnv-plus';

import { createContext as _createContext } from '../context';
import type { Fragment, IdGenerator } from '../types';
import { resolveRef } from './resolver';

const HASH_32_KINDS = ['http_operation', 'schema'];

function getKind(input: string) {
  const i = input.indexOf('-');
  return input.slice(0, i);
}

const DEFAULT_ID_GENERATOR: IdGenerator = input => {
  const kind = getKind(input);
  return HASH_32_KINDS.some(k => kind === k) ? fnv.fast1a32hex(input) : fnv.fast1a52hex(input);
};

export function createContext<T extends Fragment>(document: T) {
  return _createContext<T>(document, resolveRef, DEFAULT_ID_GENERATOR);
}
