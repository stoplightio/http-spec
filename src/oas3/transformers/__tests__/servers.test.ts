import { DeepPartial } from '@stoplight/types';
import { OpenAPIObject } from 'openapi3-ts';

import { createContext, DEFAULT_ID_GENERATOR } from '../../../context';
import { resolveRef } from '../../../oas/resolver';
import { translateToServers as _translateToServers } from '../servers';

const translateToServers = (
  document: DeepPartial<OpenAPIObject> & { paths: { '/pet': { get: Record<string, unknown> } } },
) =>
  _translateToServers.call(
    createContext(document, resolveRef, DEFAULT_ID_GENERATOR),
    document.paths['/pet'],
    document.paths['/pet'].get,
  );

describe('translateToServers', () => {
  it('translate single ServerObject to IServer', () => {
    const document = {
      servers: [
        {
          description: 'description',
          url: 'http://stoplight.io/path',
          variables: {
            a: {
              default: false,
              enum: [false, true, false],
              description: 'a - descr',
            },
            b: {
              default: 123,
              enum: [1, 2, 3],
              description: 'b - descr',
            },
          },
        },
      ],
      paths: {
        '/pet': {
          get: {},
        },
      },
    };

    expect(translateToServers(document)).toStrictEqual([
      {
        id: expect.any(String),
        description: 'description',
        url: 'http://stoplight.io/path',
        variables: {
          a: {
            default: 'false',
            description: 'a - descr',
            enum: ['false', 'true', 'false'],
          },
          b: {
            default: '123',
            description: 'b - descr',
            enum: ['1', '2', '3'],
          },
        },
      },
    ]);
  });

  it('filters out invalid variables', () => {
    const document = {
      servers: [
        {
          description: 'description',
          url: 'http://stoplight.io/path',
          variables: {
            a: null,
            b: {
              default: 123,
              enum: [1, 2, 3],
              description: 'b - descr',
            },
            c: {},
            d: {
              default: {},
            },
          } as any,
        },
      ],
      paths: {
        '/pet': {
          get: {},
        },
      },
    };

    expect(translateToServers(document)).toStrictEqual([
      {
        id: expect.any(String),
        description: 'description',
        url: 'http://stoplight.io/path',
        variables: {
          b: {
            default: '123',
            description: 'b - descr',
            enum: ['1', '2', '3'],
          },
        },
      },
    ]);
  });

  it('prefers operation servers over any other servers', () => {
    const document = {
      servers: [
        {
          description: 'description',
          url: 'http://stoplight.io/path',
        },
      ],
      paths: {
        '/pet': {
          servers: [
            {
              description: 'description',
              url: 'http://stoplight.io/pet',
            },
          ],

          get: {
            servers: [
              {
                description: 'description',
                url: 'http://stoplight.io/pet.get',
              },
            ],
          },
        },
      },
    };

    expect(translateToServers(document)).toStrictEqual([
      {
        id: expect.any(String),
        description: 'description',
        url: 'http://stoplight.io/pet.get',
      },
    ]);
  });

  it('given missing operation servers, prefers path servers over any document servers', () => {
    const document = {
      servers: [
        {
          description: 'description',
          url: 'http://stoplight.io/path',
        },
      ],
      paths: {
        '/pet': {
          servers: [
            {
              description: 'description',
              url: 'http://stoplight.io/pet',
            },
          ],

          get: {},
        },
      },
    };

    expect(translateToServers(document)).toStrictEqual([
      {
        id: expect.any(String),
        description: 'description',
        url: 'http://stoplight.io/pet',
      },
    ]);
  });
});
