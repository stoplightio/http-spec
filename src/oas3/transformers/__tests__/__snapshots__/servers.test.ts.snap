// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`translateToServers translate single ServerObject to IServer 1`] = `
Array [
  Object {
    "description": "description",
    "url": "http://stoplight.io/path",
    "variables": Object {
      "a": Object {
        "default": "false",
        "description": "a - descr",
        "enum": Array [
          "false",
          "true",
          "false",
        ],
      },
      "b": Object {
        "default": "123",
        "description": "b - descr",
        "enum": Array [
          "1",
          "2",
          "3",
        ],
      },
    },
  },
]
`;
